Added getting started and authentication page (#224)
mautic/developer-documentation-new: docs/rest_api/getting_started.rst

Getting started
###############

Mautic provides several REST API endpoints. You can find the available endpoints in the navigation menu on the left.

.. warning:: 

>    Mautic turns off the REST API by default. The Mautic administrator can turn on the API in the Mautic UI under **Configuration > API Settings**, or by setting ``'api_enabled' => 1`` in ``app/config/local.php`` directly.

Reviewer: I think this is valid and a good catch - @adiati98 for info.

>    Mautic turns off the REST API by default. The Mautic administrator can turn on the API in the Mautic UI under **Configuration > API Settings**, or by setting ``'api_enabled' => 1`` in ``app/config/local.php`` directly.

Reviewer 2: As we don't have dev docs for older versions than 5, rather than mention both locations by version, I think we better go for the correct path as below:

```suggestion
   Mautic turns off the REST API by default. The Mautic administrator can turn on the API in the Mautic UI under **Configuration > API Settings**, or by setting ``'api_enabled' => 1`` in ``config/local.php`` directly.
```

>    Mautic turns off the REST API by default. The Mautic administrator can turn on the API in the Mautic UI under **Configuration > API Settings**, or by setting ``'api_enabled' => 1`` in ``app/config/local.php`` directly.

Reviewer 3: I'd be fine to add a mention for older versions as they are still out there and alive. And there won't be any docs for older versions.

.. note:: 

    Mautic has an API library available for PHP. Check out the :xref:`Mautic API Library`.

Authentication
**************

Mautic supports Basic Authentication and OAuth2. Please see :doc:`authentication` for more details.

Error handling
**************

In case of OAuth errors, the response is a JSON encoded array similar to:

.. code-block:: json

    {
        "error": "invalid_grant",
        "error_description": "The access token provided has expired."
   }

In case of system errors, the response is a JSON encoded array similar to:

.. code-block:: json

    {
       "error": {
           "message": "You do not have access to the requested area/action.",
           "code": 403
       }
   }

> Mautic version check

Reviewer: Wrap this one in a vale off / vale on.

With the Mautic PHP API library, you can retrieve the version as follows:

.. code-block:: php

    <?php

    // Make any API request:
    $api = $this->getContext('contacts');
    $response = $api->getList('', 0, 1);

    // Get the version number from the response header:
    $version = $api->getMauticVersion();

The ``$version`` variable uses a semantic versioning format: ``[major].[minor].[patch]``, for example, ``2.4.0``. If you try it on the latest GitHub version, the version has ``-dev`` at the end, such as ``2.5.1-dev``.

API rate limiter cache
**********************

You can configure rate limiter cache in ``config/local.php``, which defaults to the filesystem:

.. code-block:: php

    <?php

    'api_rate_limiter_cache' => [ 
        'adapter' => 'cache.adapter.filesystem',
    ],

You can also configure a ``memcached`` server for improved performance:

.. code-block:: php

    <?php

    'api_rate_limiter_cache' => [
        'adapter'  => 'cache.adapter.memcached',
        'provider' => 'memcached://memcached.local:12345'
    ],

> For more examples of supported cache adapters, please visit the :xref:`Symfony Cache Documentation`.

Reviewer 3: The rate limiter was removed in M7 as the library was no longer maintained and there was no company/developer who'd care enough to find a way to keep it. If this doc is for M5 then it's OK to have it. But it should be removed for M7.

Reviewer 2: Should it also be removed for M6? We might want to create a branch for M6 so we can have all available versions.

Reviewer 3: No, it can stay in M6. It was removed in M7. Perhaps we can add a note that this exists only on versions before M7 so users would count on it?

Reviewer: Agree re. adding a note that it's been removed in 7, before we create the branch for 7.

I think there's not many changes for 6 so we should just be able to branch from 5.x (once we have the PRs merged) and leave it be, then focus on the 7 branch where there's a lot more change, as far as I know.

Getting started
###############

Mautic provides several ``REST API`` endpoints that you can use. In the navigation menu on the left, you can find the available endpoints.

.. warning:: 

    By default, Mautic's ``REST API`` isn't enabled. The Mautic administrator can enable the API in the Mautic UI under ``Configuration -> API Settings``, or by setting ``'api_enabled' => 1`` in ``app/config/local.php`` directly.

.. note:: 

>     Mautic has an API library available for PHP. You can `find it on GitHub <https://github.com/mautic/api-library>`_.

Reviewer: ```suggestion
    Mautic has an API library available for PHP. Check out the :xref:`Mautic API Library`.
```

This needs converting to a link macro. There is already a link file in /links, you can check for the name and use the xref macro as above.

To get started easily, you can use Mautic's Postman collection:

> .. image:: https://run.pstmn.io/button.svg

Reviewer: This is very incomplete right now, I'd personally leave this out until we fill it out a bit.

   :target: https://app.getpostman.com/run-collection/19345380-9b7bbddc-8a4d-437a-8fc2-42b0b9823883?action=collection%2Ffork&source=rip_markdown&collection-url=entityId%3D19345380-9b7bbddc-8a4d-437a-8fc2-42b0b9823883%26entityType%3Dcollection%26workspaceId%3D2c328b62-2531-4e35-a6bc-0f2995ce2df3

Authentication
**************

Mautic supports Basic Authentication and OAuth2. Please see :doc:`authentication` for more details.

Error handling
**************

In case of OAuth errors, the response is a JSON encoded array similar to:

.. code-block:: json

    {
        "error": "invalid_grant",
        "error_description": "The access token provided has expired."
   }

In case of system errors, the response is a JSON encoded array similar to:

.. code-block:: json

    {
       "error": {
           "message": "You do not have access to the requested area/action.",
           "code": 403
       }
   }

> .. vale off

Reviewer: Why are you turning off Vale here?

Author: Hi @RCheesley ...Thank you very much fo your feedback...it's my first time writing rst or hearing about vale....I pretty much copied the content that was linked in the PR you dropped and just made little changes, so i'm really not sure why it was initially added...hence, the reason i didn't take it off...but i'll look into it.

Mautic version check
********************

.. vale on

In case your API service wants to support several Mautic versions with different features, you might need to validate the version of Mautic you communicate with. Since Mautic 2.4.0, the version number is in all API response headers. The header name is ``Mautic-Version``.

With Mautic's PHP API library, you can get the Mautic version like this:

.. code-block:: php

    <?php

    // Make any API request:
    $api = $this->getContext('contacts');
    $response = $api->getList('', 0, 1);

    // Get the version number from the response header:
    $version = $api->getMauticVersion();

``$version`` is in a semantic versioning format: ``[major].[minor].[patch]``. For example: ``2.4.0``. If you'll try it on the latest GitHub version, the version has ``-dev`` at the end. Like ``2.5.1-dev``.

> .. vale off

Reviewer: There is no need to turn Vale off here. Instead, fix the problem :)

> API Rate limiter cache

Reviewer: ```suggestion
API rate limiter cache
```

Reviewer: Looks like this one isn't yet resolved.

Getting started with REST API
#############################

Mautic provides several REST API endpoints to manipulate and retrieve information about various entities in Mautic. You can find the available endpoints in the navigation menu on the left.

.. warning::

>    Mautic turns off the REST API by default. The Mautic administrator can turn on the API in the Mautic UI under **Configuration > API Settings**, or by setting ``'api_enabled' => 1`` in ``config/local.php`` directly. In Mautic versions earlier than 5, this setting is in ``app/config/local.php``.

Reviewer 2: I've applied the copilot suggestion here.

   Refer to the README in the GitHub repository for further instructions on using the library, or review the code examples throughout this documentation.

API rate limiter cache
**********************

.. attention::

>    The rate limiter cache is no longer available in Mautic version 7 or later.

Reviewer 2: @RCheesley is the wording here correct?

Reviewer: That's fine!
